lib/pure/unittest.nim

```diff
@@ -493,32 +534,59 @@ template suite*(name, body) {.dirty.} =
   ##  [Suite] test suite for addition
   ##    [OK] 2 + 2 = 4
   ##    [OK] (2 + -2) != 4
-  bind formatters, ensureInitialized, suiteEnded
-
-  block:
-    template setup(setupBody: untyped) {.dirty, used.} =
-      var testSetupIMPLFlag {.used.} = true
-      template testSetupIMPL: untyped {.dirty.} = setupBody
 
-    template teardown(teardownBody: untyped) {.dirty, used.} =
-      var testTeardownIMPLFlag {.used.} = true
-      template testTeardownIMPL: untyped {.dirty.} = teardownBody
-
-    let testSuiteName {.used.} = name
-
-    ensureInitialized()
-    try:
-      for formatter in formatters:
-        formatter.suiteStarted(name)
-      body
-    finally:
-      suiteEnded()
-
-proc exceptionTypeName(e: ref Exception): string {.inline.} =
-  if e == nil: "<foreign exception>"
-  else: $e.name
-
-template test*(name, body) {.dirty.} =
+  body.expectKind(nnkStmtList)
+  # parse top level suite constructs and put their structure into local variables.
+  var setup: NimNode
+  var teardown: NimNode
+  var tests: seq[NimNode]
+
+  # stuff that really should not be placed top level in a unittest.
+  result = newStmtList()
+
+  for command in body:
+    if command.kind in nnkCallKinds:
+      let typ = command[0]
+      typ.expectKind nnkIdent
+      if typ.eqIdent("test"):
+        command.expectLen 3
+        let testName = command[1]
+        testName.expectKind {nnkStrLit..nnkTripleStrLit}
+        tests.add command
+      elif typ.eqIdent("setup"):
+        command.expectLen 2
+        if setup == nil:
+          setup = command
+        else:
+          error("double definition of setup", command)
+      elif typ.eqIdent("teardown"):
+        command.expectLen 2
+        if teardown == nil:
+          teardown = command
+        else:
+          error("double definition of teardown", command)
+      else:
+        error("illegal construct for unittest suite", command)
+    else:
+      warning("please only put `test`, `setup` and `teardown` at top level of `suite`", command)
```

Comment:
In the future I think supporting more nesting would be great, needs some further adjustments, of course.